### DISCUSSION TITLE
ArgumentStringHandler blocks usage of variables

### DISCUSSION THREAD

**Discussion by makp0** [nuke-build/nuke#1164] [Q&A] [answered]
### Usage Information

7.0.0, NET 7.0

### Description

ArgumentStringHandler does not allow variables to be passed as arguments.

### Reproduction Steps

var cloneCommand = "clone https://github.com/nuke-build/nuke.git";
Git(cloneCommand);

### Expected Behavior

There have to be an overload allowing to pass strings.

### Actual Behavior

Argument is 'value' while parameter is declared as 'ref'

### Regression?

Yes. It's a completely avoidable breaking change.

### Could you help with a pull-request?

No

**Comment by matkoch** [comment]
FYI, this has been discussed here: https://github.com/nuke-build/nuke/issues/1007

From the upcoming blog post:

> As a consequence of introducing this feature, all `arguments` must be passed as interpolated strings; even if you’re invoking a tool without any arguments, you have to pass `$“”`. This has two reasons. On the one hand, interpolated string handlers cannot deal with constant strings. On the other hand, there can’t be another overload with just `string`, because interpolated strings composed of just constant strings would be handled by the `DefaultInterpolatedStringHandler`, which would render this feature unreliable.

**Reply by Droppers** [reply]
Any estimate on when this blog post will be published?

**Reply by Droppers** [reply]
@matkoch   What is the state of this blog post? We are not a fan of the argument string handler, the argument string handler only works in very simple situations, but as soon as your have to pass a lot of arguments or conditionals it becomes a pain.

I still can't see why the `Arguments` and argument string handler cannot co-exist. We wrote our own wrapper around the argument string handler to mimic the old `Arguments` class.

**Comment by matt-richardson** [comment]
I ran into trouble here too - I couldn't get it to work as expected... 

Before:
```cs
var heatArguments = $"dir {serverOutputPath} " +
    "-nologo " +
    // <snip/>
    $"-out {heatGeneratedWxsFile}";
var wiXHeatTool = ToolResolver.GetPackageTool("wix", "heat.exe");
wiXHeatTool(heatArguments);
```

After:
```cs
var wiXHeatTool = ToolResolver.GetNuGetTool("wix", "heat.exe");
var heatArguments = new ArgumentStringHandler();
heatArguments.AppendLiteral($"dir {serverOutputPath} ");
heatArguments.AppendLiteral("-nologo ");
// <snip/>
heatArguments.AppendLiteral($"-out {heatGeneratedWxsFile}");
wiXHeatTool(ref heatArguments);
```

The only thing that I could get to work (though I didn't spend _too_ long on it) was to directly call the `ArgumentStringHandler` methods to build up the args.

**Reply by matt-richardson** [reply]
Hmmm. I wonder if the string concatenation was forcing a conversion from interpolated strings to a fixed string 🤔

**Reply by matt-richardson** [reply]
Update:

I got it working via:

```cs
var wiXHeatTool = ToolResolver.GetPackageTool("wix", "heat.exe");
wiXHeatTool($"dir {serverOutputPath} " +
    $"-nologo " +
    // <snip/>
    $"-out {heatGeneratedWxsFile}");
```

Note that if you're using string concatenation (the `+`), then every part of the string needs to be an interpolated string (prefixed with `$`).

**Comment by maxcherednik** [comment, marked as answer]
Will duplicate my message here. Might be useful.

Was struggling with the upgrade from 6.3.0 to 7.0.0

After the upgrade the parameter of the node was double quoted by default.

NodeExe($"{yarnCommand}", workingDirectory: path, logger: LogErrorAsWarning);
I had to add the nq format.

NodeExe($"{yarnCommand:nq}", workingDirectory: path, logger: LogErrorAsWarning);
I guess it should have been the default option.

Am I doing something wrong?

**Reply by capjan** [reply]
@maxcherednik - thanks a lot!

I wrote the same code to convert my prepared arguments **and** also forgot to include the now required :nq formatter.

Maybe a note to the nq formatter should the added to the documentation ([Executing CLI Tools](https://nuke.build/docs/common/cli-tools/)).
That would help developers to execute their prepared arguments with quotes in the intended and expected way.
